src/crimson/os/seastore/segment_manager/block.cc

```diff
@@ -0,0 +1,416 @@
+    [=, &device](auto &bp) {
+      bufferlist bl;
+      encode(sb, bl);
+      auto iter = bl.begin();
+      assert(bl.length() < sb.block_size);
+      iter.copy(bl.length(), bp.c_str());
+      logger().debug("write_superblock: doing writeout");
+      return do_write(
+	device,
+	0,
+	bp);
+    });
+}
+
+static
+BlockSegmentManager::access_ertr::future<block_sm_superblock_t>
+read_superblock(seastar::file &device, seastar::stat_data sd)
+{
+  assert(ceph::encoded_sizeof_bounded<block_sm_superblock_t>() <
+	 sd.block_size);
+  return seastar::do_with(
+    bufferptr(ceph::buffer::create_page_aligned(sd.block_size)),
+    [=, &device](auto &bp) {
+      return do_read(
+	device,
+	0,
+	bp
+      ).safe_then([=, &bp] {
+	  bufferlist bl;
+	  bl.push_back(bp);
+	  block_sm_superblock_t ret;
+	  auto bliter = bl.cbegin();
+	  decode(ret, bliter);
+	  return BlockSegmentManager::access_ertr::future<block_sm_superblock_t>(
+	    BlockSegmentManager::access_ertr::ready_future_marker{},
+	    ret);
+      });
+    });
+}
+
+BlockSegment::BlockSegment(
+  BlockSegmentManager &manager, segment_id_t id)
+  : manager(manager), id(id) {}
+
+segment_off_t BlockSegment::get_write_capacity() const
+{
+  return manager.get_segment_size();
+}
+
+Segment::close_ertr::future<> BlockSegment::close()
+{
+  manager.segment_close(id);
+  return close_ertr::now();
+}
+
+Segment::write_ertr::future<> BlockSegment::write(
+  segment_off_t offset, ceph::bufferlist bl)
+{
+  if (offset < write_pointer || offset % manager.superblock.block_size != 0)
+    return crimson::ct_error::invarg::make();
+
+  if (offset + bl.length() > manager.superblock.segment_size)
+    return crimson::ct_error::enospc::make();
+
+  return manager.segment_write({id, offset}, bl);
+}
+
+Segment::close_ertr::future<> BlockSegmentManager::segment_close(segment_id_t id)
+{
+  assert(tracker);
+  tracker->set(id, segment_state_t::CLOSED);
+  return tracker->write_out(device, superblock.tracker_offset);
+}
+
+Segment::write_ertr::future<> BlockSegmentManager::segment_write(
+  paddr_t addr,
+  ceph::bufferlist bl,
+  bool ignore_check)
+{
+  assert((bl.length() % superblock.block_size) == 0);
+  logger().debug(
+    "segment_write to segment {} at offset {}, physical offset {}, len {}",
+    addr.segment,
+    addr.offset,
+    get_offset(addr),
+    bl.length());
+
+  
+  // TODO send an iovec and avoid the copy -- bl should have aligned
+  // constituent buffers and they will remain unmodified until the write
+  // completes
+  return seastar::do_with(
+    bufferptr(ceph::buffer::create_page_aligned(bl.length())),
+    [&](auto &bp) {
+      auto iter = bl.cbegin();
+      iter.copy(bl.length(), bp.c_str());
+      return do_write(device, get_offset(addr), bp); // TODO flushes
+    });
+}
+
+BlockSegmentManager::~BlockSegmentManager()
+{
+}
+
+BlockSegmentManager::mount_ret BlockSegmentManager::mount(mount_config_t config)
+{
+  return open_device(
+    config.path, seastar::open_flags::rw
```

Comment:
does it make sense to pass `open_flags::dsync` as well? so we don't need to bother with the `flush()` call after writing to the storage device.

Reply:
I'm not actually sure dsync ensures that a write barrier has been issued to the disk.

Reply:
i am sure. 


>        O_DSYNC
>               Write operations on the file will complete according to the
>               requirements of synchronized I/O data integrity completion.
> 
>               By the time write(2) (and similar) return, the output data has
>               been transferred to the underlying hardware, along with any
>               file metadata that would be required to retrieve that data
>               (i.e., as though each write(2) was followed by a call to
>               fdatasync(2)).  See NOTES below.
> 
> 

see also https://man7.org/linux/man-pages/man2/open.2.html

also, FWIW, seastar's implementation of `posix_file_impl::flush()` is a no-op if the file is opened with `open_flags::dsync`.

Reply:
also checked the linux kernel, it does respect the flag passed to `open()` when performing io using `file`.

see https://github.com/torvalds/linux/blob/cd796ed3345030aa1bb332fe5c793b3dddaf56e7/include/linux/fs.h#L3251 and https://github.com/torvalds/linux/blob/6ad4bf6ea1609fb539a62f10fca87ddbd53a0315/fs/block_dev.c#L199-L207